launch/pipeline/service.py

```diff
@@ -0,0 +1,76 @@
+import inspect
+from typing import Any, Callable, Dict, List, Optional
+
+from launch.pipeline.deploy import Deployment
+from launch.pipeline.runtime import Runtime
+
+
+class ServiceDescription:
+    """
+    Base ServiceDescription.
+    """
+
+    def call(self, *args, **kwargs):
+        raise NotImplementedError()
+
+
+class SingleServiceDescription(ServiceDescription):
+    """
+    The description of a service.
+    """
+
+    def __init__(
+        self,
+        service: Callable,
+        runtime: Runtime,
+        deploy: Deployment,
+        init_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__()
+        self.service = service
+        self.runtime = runtime
+        self.deploy = deploy
+        self.init_kwargs = init_kwargs
+
+        self._callable_service: Optional[Callable] = None
+
+    def _init_callable_service(self):
+        if self._callable_service:
+            return
+        if inspect.isclass(self.service):
+            kwargs = self.init_kwargs or {}
+            self._callable_service = self.service(**kwargs)
+        else:
+            assert (
+                not self.init_kwargs
+            ), "`init_kwargs` is given, but the service is not a class"
+            self._callable_service = self.service
+
+    def call(self, *args, **kwargs):
+        self._init_callable_service()
+        return self._callable_service(*args, **kwargs)
+
+
+class SeqPipelineServiceDescription(ServiceDescription):
+    """
+    The description of a sequential pipeline.
+    """
+
+    def __init__(
+        self,
+        items: List[SingleServiceDescription],
+    ):
+        super().__init__()
+        self.items = items
+
+    def call(self, *args, **kwargs):
+        for item in self.items:
+            res = item.call(*args, **kwargs)
+            if isinstance(res, tuple):
+                # Assuming call() returns multiple positional outputs
+                args = res
+            else:
+                # Handle a case when a service returns one output, so that we can pass it further using *args
+                args = (res,)
+            kwargs = {}
```

Comment:
Yes, there is an option to call `global_pipeline.call(foo=..., woo=...)` as the user knows the first step in their pipeline.
The pipeline code on the other hand, could not have information about it and can only rely on the order of the arguments and the resulting tuple

From the perspective of a user I think it makes sense to let the, use named arguments. From the API developer, yes probably we can restrict users to use only positional arguments...

Reply:
I'd simplify by restricting service's `call` methods to only accept one input parameter. Users can always achieve multi-inputs by using a `tuple` and it's not that much of a burdensome workaround. I.e. instead of 
```
def call(self, myarg1: str, myarg2: int, myarg3: float):
  ...
```
They have to do:
```
def call(self, myargs: Tuple[str, int, float]):
  myarg1, myarg2, myarg3 = *myargs
  ...
```

`Service`'s should also then return only one value.

Reply:
ok, I've made `req: Any` as a single argument. It can be a tuple or dict or tensor or anything else